### PR TITLE
enable multiple num targets

### PR DESCRIPTION
**Pull request by umertens**
This PR adds the following options:

1. training on more than one target for a particular dataframe (-> bigger training set)
2. store embedded rows within a folder named after the respective target
3. common scaling for all numerical columns (atm QuantileTransformer)
4. use torch.amp for mixed-precision training

### CODE REVIEW

File: tabgpt/col_embed.py [hunk 1]
```diff
@@ -96,38 +101,54 @@ def embed(self, n_cols, null_treatment="zero-embedding", fillna_categorical="mis
...
-                logging.info('Folder not empty, removing old files')
-                for f in files:
-                    os.remove(f)
-            for i in range(features_embeds.shape[0]):
-                file_name = f"{self.df_loader.name};{df[self.df_loader.target_column].iloc[i]};{i}"
-                np.save(os.path.join(path,file_name), features_embeds[i])
-        else:
-            return features_embeds
+        for target_col in self.df_loader.target_column:
+            descr = f'{self.df_loader.task_description} - {target_col} prediction'
```
Comment (umertens) [inline tabgpt/col_embed.py:105]: The task description should now be target-independent and a more general description of the dataframe at hand

File: tabgpt/col_embed.py [hunk 2]
```diff
@@ -96,38 +101,54 @@ def embed(self, n_cols, null_treatment="zero-embedding", fillna_categorical="mis
...
+                target_embed = self.model(**input_ids.to(device)).last_hidden_state.mean(dim=1).cpu()
+            target_embed = target_embed.repeat(len(df), 1, 1) # nrows, 1, emb_dim
+
+            features_embeds_wo_pos = torch.cat((target_embed, inputs_embeds), dim=1)
+            rows, features, emb_dim = features_embeds_wo_pos.shape
+            features_embeds = torch.zeros(rows, features, emb_dim)
+            features_embeds[:, 1:, :] = 1.
+            features_embeds += features_embeds_wo_pos
+
+            features_embeds = remove_index(features_embeds, column_identifier[target_col])
```
Comment (umertens) [inline tabgpt/col_embed.py:118]: Here the target information is removed from the features

File: tabgpt/data/house_prices/data_setup.py [hunk 3]
```diff
@@ -49,31 +50,32 @@ def setup(self, all_features=True):
         df_train_full, numerical_features, categorical_features = self.read(all_features)
        
         df_train, df_val = train_test_split(df_train_full, test_size=0.2, random_state=666)
-        num_max = df_train[numerical_features].abs().max()
-        df_train[numerical_features] = df_train[numerical_features] / num_max
-        df_val[numerical_features] = df_val[numerical_features] / num_max
-        df_train["target"] = np.log(1 + df_train["SalePrice"])
-        df_val["target"] = df_val["SalePrice"]
+       
+        self.setup_scaler(numerical_features)
```
Comment (umertens) [inline tabgpt/data/house_prices/data_setup.py:54]: You now have to set `setup_scaler` , `scale_columns` and `set_target_column` within the setup method

File: tabgpt/data/house_prices/data_setup.py [hunk 4]
```diff
@@ -38,7 +38,8 @@ def read(self, all_features=True):
                 self.old_to_new_col_mapping["GrLivArea"],
                 self.old_to_new_col_mapping["GarageArea"],
                 self.old_to_new_col_mapping["TotalBsmtSF"],
-                self.old_to_new_col_mapping["YearBuilt"],         
+                self.old_to_new_col_mapping["YearBuilt"],  
+                'SalePrice'
```
Comment (umertens) [inline tabgpt/data/house_prices/data_setup.py:42]: And always make sure to include the actual target in `numerical_features`

File: tabgpt/col_embed.py (hunk 1 above)
Comment (FelixWick) [reply tabgpt/col_embed.py:105]: But we might still want to include a target-specific task description somewhere, right?

**Comment by FelixWick** [comment]
Looks good! I like it.

File: tabgpt/col_embed.py (hunk 1 above)
Comment (umertens) [reply tabgpt/col_embed.py:105]: Sure, this happens now  [here](https://github.com/FelixWick/tabGPT/blob/multiple_targets/tabgpt/col_embed.py#L105)

**Review by FelixWick — APPROVED** [review]

**Pull request merged**